src/python/pants/backend/jvm/tasks/analysis_extraction.py

```diff
@@ -47,106 +46,89 @@ def _create_products_if_should_run(self):
     # If none of our computed products are necessary, return immediately.
     if not self._create_products_if_should_run():
       return
 
-    zinc_analysis = self.context.products.get_data('zinc_analysis')
     classpath_product = self.context.products.get_data('runtime_classpath')
-    classes_by_source = self.context.products.get_data('classes_by_source')
-    product_deps_by_src = self.context.products.get_data('product_deps_by_src')
+    product_deps_by_target = self.context.products.get_data('product_deps_by_target')
 
     fingerprint_strategy = DependencyContext.global_instance().create_fingerprint_strategy(
         classpath_product)
 
-    targets = list(zinc_analysis.keys())
-    with self.invalidated(targets,
-                          fingerprint_strategy=fingerprint_strategy,
-                          invalidate_dependents=True) as invalidation_check:
-      # Extract and parse products for any relevant targets.
-      for vt in invalidation_check.all_vts:
-        summary_json_file = self._summary_json_file(vt)
-        cp_entry, _, analysis_file = zinc_analysis[vt.target]
-        if not vt.valid:
-          self._extract_analysis(vt.target, analysis_file, summary_json_file)
-        self._register_products(vt.target,
-                                cp_entry,
-                                summary_json_file,
-                                classes_by_source,
-                                product_deps_by_src)
-
-  def _extract_analysis(self, target, analysis_file, summary_json_file):
-    target_classpath = self._zinc.compile_classpath('runtime_classpath', target)
-    analysis_by_cp_entry = self._analysis_by_runtime_entry
-    upstream_analysis = list(self._upstream_analysis(target_classpath, analysis_by_cp_entry))
-    args = [
-        '-summary-json', summary_json_file,
-        '-analysis-cache', analysis_file,
-        '-classpath', ':'.join(target_classpath),
-        '-analysis-map', ','.join('{}:{}'.format(k, v) for k, v in upstream_analysis),
-      ]
-
-    result = self.runjava(classpath=self._zinc.extractor,
-                          main=Zinc.ZINC_EXTRACT_MAIN,
-                          args=args,
-                          workunit_name=Zinc.ZINC_EXTRACTOR_TOOL_NAME,
-                          workunit_labels=[WorkUnitLabel.MULTITOOL])
-    if result != 0:
-      raise TaskError('Failed to parse analysis for {}'.format(target.address.spec),
-                      exit_code=result)
-
-  def _upstream_analysis(self, target_classpath, analysis_by_cp_entry):
-    for entry in target_classpath:
-      analysis_file = analysis_by_cp_entry.get(entry)
-      if analysis_file is not None:
-        yield entry, analysis_file
+    targets = self.context.targets()
+    # complete classpath list
+    deps_classpaths = [
+      cp[1] for target in targets for cp in classpath_product.get_for_target(target) if cp[1]
+    ]
+    with self.aliased_classpaths(deps_classpaths) as deps_classpath_by_alias:
```

Comment:
While I can understand doing this outside of the invalidated block, I think that it would probably be more self contained to do it inside of the `def _run_jdeps_analysis` method. Yes, it would re-run for each analyzed target, but invalidated blocks are easy enough to get wrong that I recommend keeping them very, very simple.